File: continuum_context_inject.py
#!/usr/bin/env python3
"""UserPromptSubmit hook: auto-inject relevant Continuum memory into every
message, without relying on the model deciding to call memory_search.

Installed by Continuum's install-hook script. Pure stdlib and no shell, so the
same file runs unchanged on Linux, macOS and Windows. Never blocks the prompt:
any failure, timeout, or missing token exits 0 silently and the message
proceeds with no injected context.
"""

import json
import os
import sys
import urllib.request
from pathlib import Path

STATE_DIR = Path.home() / ".continuum"
TOKEN_FILE = STATE_DIR / "hook-token"
DISABLE_FILE = STATE_DIR / "hook-disabled"
WORKSPACE_MAP = STATE_DIR / "workspace-map.json"

CONTINUUM_URL = os.environ.get("CONTINUUM_MCP_URL", "https://continuum-mcp.sshogunn.org")
TIMEOUT_SECONDS = 3.0
RECENT_TURNS = 6
RECENT_TURN_CHARS = 600
RECENT_TOTAL_CHARS = 2000


def text_of(message):
    content = message.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return " ".join(
            block.get("text", "")
            for block in content
            if isinstance(block, dict) and block.get("type") == "text"
        )
    return ""


def recent_turns(transcript_path):
    if not transcript_path:
        return ""
    turns = []
    try:
        with open(transcript_path, encoding="utf-8", errors="replace") as f:
            for line in f:
                try:
                    entry = json.loads(line)
                except ValueError:
                    continue
                message = entry.get("message")
                if not isinstance(message, dict):
                    continue
                role = message.get("role")
                if role not in ("user", "assistant"):
                    continue
                body = text_of(message).strip()
                if body:
                    turns.append("%s: %s" % (role, body[:RECENT_TURN_CHARS]))
    except OSError:
        return ""
    return "\n".join(turns[-RECENT_TURNS:])[-RECENT_TOTAL_CHARS:]


def workspace_for(cwd):
    if not cwd:
        return ""
    try:
        with open(WORKSPACE_MAP, encoding="utf-8") as f:
            return json.load(f).get(cwd, "") or ""
    except (OSError, ValueError):
        return ""


def post_context(token, payload):
    request = urllib.request.Request(
        CONTINUUM_URL.rstrip("/") + "/hook/context",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "Authorization": "Bearer " + token,
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=TIMEOUT_SECONDS) as response:
        return json.loads(response.read().decode("utf-8"))


def main():
    if DISABLE_FILE.exists():
        return
    try:
        token = TOKEN_FILE.read_text(encoding="utf-8").strip()
    except OSError:
        return
    if not token:
        return

    data = json.load(sys.stdin)
    prompt = data.get("prompt") or ""
    if not prompt:
        return

    payload = {"query": prompt}
    workspace = workspace_for(data.get("cwd") or "")
    if workspace:
        payload["workspace"] = workspace
    recent = recent_turns(data.get("transcript_path") or "")
    if recent:
        payload["recent"] = recent

    context = post_context(token, payload).get("context")
    if context:
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "UserPromptSubmit",
                "additionalContext": context,
            }
        }))


if __name__ == "__main__":
    for stream in (sys.stdin, sys.stdout):
        try:
            stream.reconfigure(encoding="utf-8", errors="replace")
        except (AttributeError, ValueError):
            pass
    try:
        main()
    except Exception:
        pass
    sys.exit(0)

File: hooks/continuum_context_inject.js
#!/usr/bin/env node
"use strict";

// UserPromptSubmit hook: auto-inject relevant Continuum memory into every
// message, without relying on the model deciding to call memory_search.
//
// Installed by Continuum's install-hook script. Pure Node core modules and no
// dependencies, so the same file runs unchanged on Linux, macOS and Windows.
// Never blocks the prompt: any failure, timeout, or missing token exits 0
// silently and the message proceeds with no injected context.

const fs = require("fs");
const os = require("os");
const path = require("path");
const { spawn } = require("child_process");

const STATE_DIR = path.join(os.homedir(), ".continuum");
const TOKEN_FILE = path.join(STATE_DIR, "hook-token");
const DISABLE_FILE = path.join(STATE_DIR, "hook-disabled");
const WORKSPACE_MAP = path.join(STATE_DIR, "workspace-map.json");
const UPDATE_STAMP = path.join(STATE_DIR, "update-checked-at");
const UPDATE_DISABLE_FILE = path.join(STATE_DIR, "no-auto-update");
const UPDATER = path.join(os.homedir(), ".claude", "hooks", "continuum-self-update.js");

const CONTINUUM_URL = process.env.CONTINUUM_MCP_URL || "https://continuum-mcp.sshogunn.org";
const TIMEOUT_MS = 3000;
const RECENT_TURNS = 6;
const RECENT_TURN_CHARS = 600;
const RECENT_TOTAL_CHARS = 2000;

function spawnUpdate() {
  if (fs.existsSync(UPDATE_DISABLE_FILE) || !fs.existsSync(UPDATER)) return;
  const hours = parseFloat(process.env.CONTINUUM_UPDATE_INTERVAL_HOURS || "24") || 24;
  try {
    if (Date.now() - fs.statSync(UPDATE_STAMP).mtimeMs < hours * 3600 * 1000) return;
  } catch {
    // no stamp yet — proceed
  }
  fs.mkdirSync(STATE_DIR, { recursive: true });
  fs.writeFileSync(UPDATE_STAMP, String(Date.now()), "utf-8");

  const child = spawn(process.execPath, [UPDATER], { detached: true, stdio: "ignore" });
  child.unref();
}

function textOf(message) {
  const content = message.content;
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .filter((block) => block && typeof block === "object" && block.type === "text")
      .map((block) => block.text || "")
      .join(" ");
  }
  return "";
}

function recentTurns(transcriptPath) {
  if (!transcriptPath) return "";
  let raw;
  try {
    raw = fs.readFileSync(transcriptPath, "utf-8");
  } catch {
    return "";
  }
  const turns = [];
  for (const line of raw.split("\n")) {
    if (!line.trim()) continue;
    let entry;
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }
    const message = entry.message;
    if (!message || typeof message !== "object") continue;
    const role = message.role;
    if (role !== "user" && role !== "assistant") continue;
    const body = textOf(message).trim();
    if (body) turns.push(`${role}: ${body.slice(0, RECENT_TURN_CHARS)}`);
  }
  return turns.slice(-RECENT_TURNS).join("\n").slice(-RECENT_TOTAL_CHARS);
}

function workspaceFor(cwd) {
  if (!cwd) return "";
  try {
    return JSON.parse(fs.readFileSync(WORKSPACE_MAP, "utf-8"))[cwd] || "";
  } catch {
    return "";
  }
}

async function postContext(token, payload) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const response = await fetch(`${CONTINUUM_URL.replace(/\/$/, "")}/hook/context`, {
      method: "POST",
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    return await response.json();
  } finally {
    clearTimeout(timer);
  }
}

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf-8");
    process.stdin.on("data", (chunk) => (data += chunk));
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
  });
}

async function main() {
  if (fs.existsSync(DISABLE_FILE)) return;
  let token;
  try {
    token = fs.readFileSync(TOKEN_FILE, "utf-8").trim();
  } catch {
    return;
  }
  if (!token) return;

  let data;
  try {
    data = JSON.parse(await readStdin());
  } catch {
    return;
  }
  const prompt = data.prompt || "";
  if (!prompt) return;

  const payload = { query: prompt };
  const workspace = workspaceFor(data.cwd || "");
  if (workspace) payload.workspace = workspace;
  const recent = recentTurns(data.transcript_path || "");
  if (recent) payload.recent = recent;

  const result = await postContext(token, payload);
  const context = result && result.context;
  if (context) {
    process.stdout.write(
      JSON.stringify({
        hookSpecificOutput: { hookEventName: "UserPromptSubmit", additionalContext: context },
      }) + "\n"
    );
  }
}

(async () => {
  try {
    await main();
  } catch {
    // never block the prompt
  }
  try {
    spawnUpdate();
  } catch {
    // best effort
  }
  process.exit(0);
})();
